ios/Runner/AppAuthHandler.swift

```diff
@@ -0,0 +1,304 @@
+    
+    init(methodChannel: FlutterMethodChannel, controller: UIViewController) {
+        self.authMethodChannel = methodChannel
+        self.controller = controller
+        os_log("Init App Auth Handler")
+    }
+    
+    private func invokeCompletionHandler(tokens: Tokens) {
+        while !self.completionHandler.isEmpty {
+            self.completionHandler.removeFirst()(tokens)
+        }
+    }
+    
+    private func invokeErrorHandler(errorMessage : String) {
+        while !self.errorHandler.isEmpty {
+            self.errorHandler.removeFirst()(errorMessage)
+        }
+    }
+    
+    /**
+     Gets a token from auth
+     */
+    func getAccessTokens(_ session : Session, completionHandler : @escaping (Tokens) -> Void, errorHandler:@escaping (String) -> Void) {
+        self.currentSession = session
+        
+        self.completionHandler.append(completionHandler)
+        self.errorHandler.append(errorHandler)
+        
+        let state = session.state // Provided session has a state in textform, need to get a valid state from userSettings
+        
+        // A incomming session could be searche internally for existing tokens
+        
+        if (state == nil) {
+            os_log("Provided session for fetching access token has no authorization, launching authorization process")
+            authorizeUser()
+        } else {
+            
+            os_log("State was set previously, refreshing token from State..")
+            
+            // A textual state was set, recover laste stored State Obejct
+            recoverToken()
+        }
+    }
+    
+    func recoverToken() {
+        guard let session = self.currentSession else {
+            os_log("Error in getting last session")
+            self.errorHandler.first?("Error in getting last session")
+            return
+        }
+        
+        if let authState = loadState(forUserId: session.userId) {
+            // State exists, perform refreshing
+            
+            if (!authState.isAuthorized) {
+                authorizeUser()
+                return
+            }
+            
+            guard let tokenResponse = authState.lastTokenResponse else {
+                self.invokeErrorHandler(errorMessage: "No token received")
+                return
+            }
+            let tokens = Tokens(accessToken: tokenResponse.accessToken!, idToken: tokenResponse.idToken!)
+            self.invokeCompletionHandler(tokens: tokens)
+            
+        } else {
+            // State do not exist, reauthenticate
+            authorizeUser()
+        }
+    }
+    
+    func authorizeUser() {
+        guard let userId = self.currentSession?.userId else {
+            os_log("No userID set.")
+            self.invokeErrorHandler(errorMessage: "No user ID set")
+            return
+        }
+        
+        // Get registration from device store
+        let registration = RegistrationStore.fetchRegistration(userId: userId)
```

Comment:
We should make sure to reuse an existing application registration to the same host, even when it was for another user, rather than spamming pods with many registrations for our app :) 

Given you're only using the hostname from the stored registration, I'm not sure that happens?

For Android I have chosen to proxy the actual storage of registrations back into the Dart engine, to make use of the platform independent persistence abstraction: 

https://github.com/jhass/insporation/blob/aab1d23f80472c3a9dabd9adcadd5d676da44a89/android/app/src/main/kotlin/jhass/eu/insporation/AppAuthHandler.kt#L272-L289

https://github.com/jhass/insporation/blob/aab1d23f80472c3a9dabd9adcadd5d676da44a89/lib/src/app_auth.dart#L118-L141

Then the stored state, if found, is reinstantiated instead of performing a registration each time:

https://github.com/jhass/insporation/blob/main/android/app/src/main/kotlin/jhass/eu/insporation/AppAuthHandler.kt#L169

https://github.com/jhass/insporation/blob/main/android/app/src/main/kotlin/jhass/eu/insporation/AppAuthHandler.kt#L328 

`AuthState` is a container by the Android lib to store various state objects, namely the registration and tokens.

Reply:
Would you mind refactor the dart part of 'fetchRegistration' from 'userId' parameter to 'hostName'? Current implementation look a bit misleading, if not diving down till the real implementation.

Reply:
Not too much. I think my idea was to centralize the logic for parsing the hostname from the userId a bit rather than pushing it to all platform implementations.